File: q2140/q2140_recursive.java
class Solution {
    public long mostPoints(int[][] questions) {
        long[] memo = new long[questions.length];
        Arrays.fill(memo, -1);
        return dpHelper(questions, 0, memo); 
    }

    // dpHelper(i) represents the maxPoints you could accumulate if you start answering questions from the i-th question 
    private long dpHelper(int[][] questions, int i, long[] memo) {
        // base case: if we are already at the last one return 0
        if (i >= questions.length) return 0;

        // memoized case: 
        if (memo[i] != -1) return memo[i];

        // recursive case: 
        int[] currQuestion = questions[i];
        long skip = dpHelper(questions, i+1, memo);
        long noSkip = currQuestion[0] + dpHelper(questions, i + currQuestion[1] + 1, memo);
        
        // memoize the result
        memo[i] = Math.max(skip, noSkip);
        return memo[i];
    }
}

File: q2140/q2140_recursive.go
func mostPoints(questions [][]int) int64 {
    memo := make([] int64, len(questions))
    for i := range(memo) {
        memo[i] = -1
    }
    return dpHelper(0, memo, questions)
}

// represent the max points if you start from index i
func dpHelper(i int, memo []int64, questions [][]int) int64 {
    if i >= len(questions) {
        return 0
    }

    if memo[i] != -1 {
        return memo[i]
    }

    currQuestion := questions[i]
    skip := dpHelper(i+1, memo, questions)
    noSkip := int64(currQuestion[0]) + dpHelper(i + currQuestion[1] + 1, memo, questions)

    if skip > noSkip {
        memo[i] = skip
    } else {
        memo[i] = noSkip
    }
    
    return memo[i]
}
